Task data

matikkk231 opened the pull request.

matikkk231 committed:
ref: loading configs

matikkk231 committed:
add: saveSystem

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Data/BonusData.cs:
```diff
@@ -0,0 +1,47 @@
+{
+    public class BonusData : IBonusData
+    {
+        [JsonProperty("BonusLevelData")] public int BonusLevelData;
```
You have class that calls Bonus**Data**. You dont need to add **Data** to each field. You already in the context of data class. 

But another thing: you don't need to create for each property field for serailization. You can add attribute for property. Please remove all fields and use only property, because now it's overhead

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Data/BonusData.cs:
```diff
@@ -0,0 +1,47 @@
+        public int BonusLevel
+        {
+            get => BonusLevelData;
+            set => BonusLevelData = value;
```
[JsonProperty("BonusLevel")] public int BonusLevel { get; set; }

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Data/BonusData.cs:
```diff
@@ -0,0 +1,47 @@
+            UpgradeValueData = 0;
+            ProvidingDamagePerTapBonusData = 0;
+            IdData = "Nothing";
+        }
```
You don't need constructor to set default values. For string default value is null, that why you had an exception of null. To solve it, you need to set the value not in constructor, you have set default value from model that use this data. Only model will change this data.

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Data/BonusData.cs:
```diff
@@ -0,0 +1,47 @@
+            set => IdData = value;
+        }
+
+        public BonusData()
```
Just to note. You have to have default constructor like you write here `public BonusData()` (you know, if you don't have any constructor you will have everytime default constructor) because serialization system works in such steps:

1) Create object of class like `new BonusData()`
2) Set values to each property or field that is public or has attribute `[JsonProperty]`

If you will not have default constructor and only constructor with parameters (for example `public BonusData(int someInt)`) the serialization system will not understand how to deserialize data...

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/BonusModel.cs:
```diff
@@ -22,14 +23,16 @@ public int ProvidingDamagePerTapBonus
                 _providingDamagePerTapBonus = value;
                 Updated?.Invoke();
-                DamagePerTapBonusChanged?.Invoke();
+                DamagePerTapBonusChanged?.Invoke(0);
```
why it invoke with 0?

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/BonusModel.cs:
```diff
@@ -57,6 +60,17 @@ public BonusModel(IBonusConfig config)
+
+            ProvidingDamagePerTapBonus = data.ProvidingDamagePerTapBonus;
+
+            _id = data.Id;
```
and what if data don't contains id? Like null? So, you have set here in this case to data the id from config.

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/BonusModel.cs:
```diff
@@ -57,6 +60,17 @@ public BonusModel(IBonusConfig config)
+
+            UpgradeValue = data.UpgradeValue;
+
+            ProvidingDamagePerTapBonus = data.ProvidingDamagePerTapBonus;
```
You never update data...

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/BonusModel.cs:
```diff
@@ -22,14 +23,16 @@ public int ProvidingDamagePerTapBonus
             }
         }
 
+        public string Id => _id;
```
public string Id => _data.Id;

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/BonusModel.cs:
```diff
@@ -22,14 +23,16 @@ public int ProvidingDamagePerTapBonus
             {
                 _providingDamagePerTapBonus = value;
                 Updated?.Invoke();
-                DamagePerTapBonusChanged?.Invoke();
```
It's looks, like you can remove _providingDamagePerTapBonus and use _data.ProvidingDamagePerTapBonus = value

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesId/BonusesId.cs:
```diff
@@ -0,0 +1,12 @@
+namespace Project.Scripts.Game.BonusesId
+{
+    public class BonusesId : IBonusesId
```
For what this class relate to? 

You already has the Id in configs but you duplicate it in this class. It's not correct to keep same data in few classes. You should say to some model that need this id, to get it from config.

vovasazonov replied:
Model don't have to know is it sword or something another. The rules should be configurable. You have to write the rules in config. So, if game designers will be say that sword will be changed to another resource you have to be ready to it just to change via config in json

vovasazonov replied:
So in short. Access to specific field id via config interface. Don't make another classes for it

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesShop/Data/BonusesShopData.cs:
```diff
@@ -0,0 +1,17 @@
+    {
+        [JsonProperty("SwordData")] public BonusData SwordData;
+
+        public IBonusData Sword => SwordData;
```
Here is okay that you did it like this. Because serialization don't know how to works with interfaces

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesShop/Model/BonusesShopModel.cs:
```diff
@@ -2,41 +2,56 @@
             AddListeners();
         }
 
         private void OnUpgradeBought(string upgradedBonus)
```
upgradedBonusId

vovasazonov replied:
or bonusId

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesShop/Model/BonusesShopModel.cs:
```diff
@@ -49,14 +64,12 @@ private void UpdateBonusInfo(string upgradedBonus)
         }
 
-        private void UpdateDamagePerTapBonus()
+        private void UpdateDamagePerTapBonus(int increaseDamagePerTapBonus)
```
you can just write bonusId

matikkk231 replied:
it isn't just a bonus, its a result after some logic which means how much damagePerTapBonus was increased

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Data/IBonusData.cs:
```diff
@@ -0,0 +1,10 @@
+    {
+        public int BonusLevel { get; set; }
+        public int UpgradeValue { get; set; }
+        public int ProvidingDamagePerTapBonus { get; set; }
```
you dont need here Bonus in name. `ProvidingDamagePerTap`

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Bonus/Model/IBonusModel.cs:
```diff
@@ -5,13 +5,13 @@ namespace Project.Scripts.Game.Areas.Bonus.Model
-        event Action<string> UpgradeBought ;
-        event Action DamagePerTapBonusChanged;
+        event Action<string> UpgradeBought;
+        event Action<int> DamagePerTapBonusChanged;
```
You can remove Bonus from name. `DamagePerTapChanged`

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesShop/Data/IBonusesShopData.cs:
```diff
@@ -0,0 +1,9 @@
+
+namespace Project.Scripts.Game.Areas.BonusesShop.Data
+{
+    public interface IBonusesShopData
```
So you can remove this interface and use config interface that you already has

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/BonusesShop/Presenter/BonusesShopPresenter.cs:
```diff
@@ -18,6 +19,13 @@ public class BonusesShopPresenter : IDisposable
+            IBonusesShopData data)
+        {
+            _bonusPresenters.Add(data.Sword.Id,
+                new BonusPresenter(bonusesShopView.Sword, bonusesShopModel.Collection[data.Sword.Id]));
```
please make your rider ability to write code up to 200 lines.

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResource/Data/GameResourceData.cs:
```diff
@@ -0,0 +1,30 @@
+    [Serializable]
+    public class GameResourceData : IGameResourceData
+    {
+        [JsonProperty("AmountData")] public int _amount;
```
You are using prefix for public fields.... It's not correct. Anyway you can remove this fields and use properties like I described above

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResource/Data/GameResourceData.cs:
```diff
@@ -0,0 +1,30 @@
+        {
+            _amount = 0;
+            _id = "Nothing";
+        }
```
This constructor can be removed

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResources/Data/GameResourcesData.cs:
```diff
@@ -0,0 +1,20 @@
+
+namespace Project.Scripts.Game.Areas.GameResources.Data
+{
+    public class GameResourcesData : IGameResourcesData
```
Good

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResources/Model/GameResourcesModel.cs:
```diff
@@ -19,5 +20,16 @@ public GameResourcesModel(IGameResourcesConfig gameResourcesConfig)
+
+            IGameResourceModel damagePerTap = new GameResourceModel(data.DamagePerTap.Id,
+                data.DamagePerTap.Amount);
+            Collection[damagePerTap.Id] = damagePerTap;
```
You wrote duplicated code but for diffrent resources...

Please create models by go throw collection of resources in data or configs...

Something like 

```
foreach(resourceData in data.Resources)
{
    var model = new GameResourceModel(resourceData );
    Collection.Add(resourceData.Id, model);
}
```

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResources/Presenter/GameResourcesPresenter.cs:
```diff
@@ -18,6 +20,16 @@ public class GameResourcesPresenter : IDisposable
+            _gameResourcePresenters.Add(new GameResourcePresenter(view.Money,
+                model.Collection[data.Money.Id]));
+            _gameResourcePresenters.Add(new GameResourcePresenter(view.DamagePerTap,
+                model.Collection[data.DamagePerTap.Id]));
```
same. Create presenters via foreach. You have contains collection of resources in  IGameResourcesModel

matikkk231 replied:
and what i supposed to with views?

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResources/Presenter/GameResourcesPresenter.cs:
```diff
@@ -18,6 +20,16 @@ public class GameResourcesPresenter : IDisposable
                 model.Collection[gameResourcesConfig.Money.Id]));
             _gameResourcePresenters.Add(new GameResourcePresenter(view.DamagePerTap,
                 model.Collection[gameResourcesConfig.DamagePerTap.Id]));
+            _data = new GameResourcesData();
```
model have to be related to create new data. It's not responsibility of presenter

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResources/Presenter/GameResourcesPresenter.cs:
```diff
@@ -10,6 +11,7 @@ namespace Project.Scripts.Game.Areas.GameResources.Presenter
     public class GameResourcesPresenter : IDisposable
     {
         private readonly List<GameResourcePresenter> _gameResourcePresenters = new();
+        private IGameResourcesData _data;
```
remove data from presenter. Only model should access to data

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/GameResourcesId/GameResourcesId.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Project.Scripts.Game.Areas.GameResourcesId
+{
+    public class GameResourcesId : IGameResourcesId
```
Same. you can access to id via config

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/LevelSystem/Data/LevelSystemData.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class LevelSystemData : ILevelSystemData
+    {
+        [JsonProperty("CurrentLevelData")] public int CurrentLevelValue;
```
remove fields and use only properties

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/LevelSystem/Data/LevelSystemData.cs:
```diff
@@ -0,0 +1,38 @@
+        {
+            CurrentLevel = 0;
+            CurrentExperience = 0;
+            ExperienceBeforeLevelUp = 0;
```
remove constructor. Whole fields has default values

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/LevelSystem/Model/LevelSystemModel.cs:
```diff
@@ -32,6 +33,13 @@ public LevelSystemModel()
 
+        public LevelSystemModel(ILevelSystemData data)
+        {
+            CurrentExperience = data.CurrentExperience;
```
please move logic of change data to model.

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/Monster/Presenter/MonsterPresenter.cs:
```diff
@@ -9,16 +10,27 @@ public class MonsterPresenter : IDisposable
     {
         private readonly IMonsterView _view;
         private readonly IMonsterModel _model;
+        private IMonsterData _data;
```
remove data from presenter. Only model should create and manupulate with data

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/SaveSystem/SaveSystem.cs:
```diff
@@ -0,0 +1,49 @@
+
+namespace Project.Scripts.Game.Areas.SaveSystem
+{
+    public static class SaveSystem
```
Please don't make it static. Create interface and implement this interface. 

Second - rename it to service instead system and move to core path.

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/Areas/SaveSystem/SaveSystem.cs:
```diff
@@ -0,0 +1,49 @@
+            fileStream.Close();
+        }
+
+        public static GameData LoadData()
```
Please make it more generic. Currently you depence to specific GameData. Make it generic.
TData LoadData<TData>()

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Game/EntryPoint.cs:
```diff
@@ -20,24 +22,72 @@ public class EntryPoint : MonoBehaviour
+            return new GameConfigs(bonusesShopConfig, gameResourcesConfig, monsterConfig);
+        }
+
+        private void OverwriteGameData()
```
You can remove this method after move the code to models

matikkk231 committed:
ref: now model manipulates with data

matikkk231 committed:
ref: fixed previos commit

matikkk231 committed:
ref: bonusesShop

now bonusView intantiate in runTime and all bonus's data and configs encapsulated

matikkk231 committed:
ref: data

matikkk231 committed:
ref: whole program

vovasazonov commented on MVP_Clicker/Assets/Project/Scripts/Core/SaveSystemService/Model/ISaveSystemService.cs:
```diff
@@ -0,0 +1,10 @@
+    public interface ISaveSystemService
+    {
+        public void SaveData(IGameData data);
+        public IGameData LoadData();
```
This service have to be able to move to other games. So it should be more generic like `void SaveData<T>(T data);`
